File: android/app/src/main/java/com/laneshadow/ui/atoms/LSBestBadge.kt
package com.laneshadow.ui.atoms

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.padding
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import com.laneshadow.theme.LocalLaneShadowTheme
import com.laneshadow.theme.generated.LaneShadowTheme.IconName
import com.laneshadow.theme.generated.LaneShadowTheme as GeneratedTokens

private const val BestBadgeLabel = "BEST FOR TODAY"

@Composable
fun LSBestBadge(modifier: Modifier = Modifier) {
    val theme = LocalLaneShadowTheme.current

    LSPill(
        size = PillSize.Sm,
        modifier = modifier.badgeSurface(
            theme = theme,
            backgroundColor = GeneratedTokens.color.Signal.default,
            foregroundColor = theme.content.onSignal,
        ),
    ) {
        Row(
            modifier = Modifier.padding(horizontal = theme.space.xs),
            horizontalArrangement = Arrangement.spacedBy(theme.space.xs),
        ) {
            LSIcon(
                name = IconName.StarFill,
                size = IconSize.Xs,
                color = IconColor.Signal,
            )

            Text(
                text = BestBadgeLabel,
                style = TypographyVariant.Ui.Label.Sm.resolveTextStyle(theme),
                color = theme.content.onSignal,
            )
        }
    }
}

File: ios/LaneShadow/Views/Atoms/LSBestBadge.swift
import LaneShadowTheme
import SwiftUI

// MARK: - Animation Motion Extensions

extension Animation {
    /// Best badge entrance animation: 200ms spring
    ///
    /// Reads from theme.motion.bestBadgeEnter (uses "standard" duration = 240ms and "emphasized" easing)
    static func bestBadgeEnter(theme: Theme) -> Animation {
        let duration = Double(theme.motion.duration["standard"] ?? 240) / 1000
        let easing = theme.motion.easing["emphasized"] ?? [0.2, 0.0, 0.0, 1.0]
        return Animation.timingCurve(
            easing[0],
            easing[1],
            easing[2],
            easing[3],
            duration: duration
        )
    }
}

public struct LSBestBadgeResolvedStyle: Equatable, Sendable {
    let backgroundToken: String
    let foregroundToken: String
    let backgroundColor: Color
    let foregroundColor: Color
    let leadingIcon: IconName
    let iconSize: IconSize

    func pillHeight(in theme: Theme) -> CGFloat {
        PillSize.sm.height(in: theme)
    }
}

public struct LSBestBadge: View {
    @Environment(\.theme) private var theme

    static let labelText = "BEST FOR TODAY"

    @State private var isAppeared = false

    public init() {}

    public var body: some View {
        let style = Self.resolvedStyle(in: theme)

        LSPill(size: .sm) {
            HStack(spacing: theme.space.xs) {
                LSIcon(
                    name: style.leadingIcon,
                    size: style.iconSize,
                    resolvedColorOverride: style.foregroundColor
                )

                Text(Self.labelText)
                    .font(theme.type.label.sm.font)
                    .foregroundStyle(style.foregroundColor)
                    .textCase(.uppercase)
            }
            .padding(.horizontal, theme.space.xs)
            .background(style.backgroundColor, in: Capsule(style: .continuous))
        }
        .scaleEffect(isAppeared ? 1.0 : 0.8)
        .opacity(isAppeared ? 1.0 : 0.0)
        .animation(Animation.bestBadgeEnter(theme: theme), value: isAppeared)
        .onAppear {
            isAppeared = true
        }
        .accessibilityElement(children: .combine)
        .accessibilityLabel(Self.labelText)
    }
}

extension LSBestBadge {
    static func resolvedStyle(in theme: Theme) -> LSBestBadgeResolvedStyle {
        LSBestBadgeResolvedStyle(
            backgroundToken: "color.signal.default",
            foregroundToken: "color.content.onSignal",
            backgroundColor: theme.colors.primary.default,
            foregroundColor: ContentColor.onSignal.resolved(in: theme),
            leadingIcon: .starFill,
            iconSize: .xs
        )
    }
}
